rfcs/002 - Compiler Refactor RFC.md

```diff
@@ -0,0 +1,141 @@
+8. take each pair of outputs from (5) and (6) and run them through a Python linter to confirm they are valid python (most stylistic rules removed, just syntax, unused imports, and a few others), if it errors it sends all of this to an LLM to generate new versions of the file, repeating on error until a limit is reached, then exiting the flow for that particular pair, erroring if all pairs fail.
+9. take each pair of outputs from (8) and runs the test suite, passing them and the test suite errors to an LLM if they fail, otherwise continuing. If it continues to fail for too long, the particular pair exits the flow, if all pairs exit the flow, it errors.
+10. the first pair to finish (9) cancels the rest of the operations, then by default it appends the CLI and HTTP server helper logic to the code file and exits.
+
+For now, *most* of the Marsha language functionality exists within (3), and it is a synchronous operation, but most of the solutions for database support would at minimum require (3) to become async and involve various DB and ML operations (likely: connect to the DB and generate a schema dump, then chunk that dump and index it with vectors, then use vector search to find the most relevant table(s), stored procedure(s), etc to provide to (5) and (6)) but it's possible these features could "bleed" into the validation steps of Marsha, and the codebase starts turning (even more) into a big ball of mud. (For instance, validating the generated SQL DB access code in (5) against the actual DB schema in (7) seems likely).
+
+There are other dimensions we would like to extend Marsha on:
+
+* Which LLM is used to do the work.
+* Which target language Marsha compiles to.
+* Dev tools (debugging, decompiling, etc) for Marsha.
+
+The solution we come up with needs to keep the complexity of all of these different features from compounding on one another, ideally.
+
+#### Proposal: Marsha as Extensions that append to lists of layers
+
+Taking a very rough inspiration from LLMs themselves, we'll have a singular kind of "transformer" that is the foundational piece of the compiler, and they will be organized into layers where each layer executes in parallel and their outputs are fed into the next layer, etc. Each logical feature will be an extension that concatenates its logic into the appropriate layers, with the "meaning" said layers being a convention per extension "group".
+
+Now working backwards, an Extension Group is a collection of extensions that all declare membership of a particular group name. Extensions won't be allowed to mix between groups to reduce the complexity involved, and because an Extension Group declares how many layers exist and their input and output formats (most layers will likely be lists of some class type, but the input for the first layer will be `None` and use the quasi-global configuration data to set itself up and the output of the last layer can only be `None` or `str`).
```

Comment:
Why not just vocabulary relating an extension to tree theory or lineage? So a group defines a tree structure (which could be implemented as an interface), a layer represents the tree level. An extension group is a group of trees that implement or follow a given tree structure.

Reply:
Well, it's not exactly a tree structure because each later layer can consume any/all of the output of the prior layer. It's a kind of acyclic directed graph but a little bit more constrained than that since you can't have branches of the graph that have one or more nodes hidden from other nodes the same "distance" from the starting node.

I don't know what the proper CS terminology for something like that.

We could go with something like this, but I also don't want the name to be misleading to people reading it (like how React misappropriated MVC terminology for different concepts).